Mathlib/RingTheory/Coprime/Basic.lean

```diff
@@ -0,0 +1,391 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Ken Lee, Chris Hughes
+-/
+import Mathlib.Tactic.Ring
+import Mathlib.GroupTheory.GroupAction.Units
+import Mathlib.Algebra.Ring.Divisibility
+import Mathlib.Algebra.Hom.Ring
+import Mathlib.Algebra.GroupPower.Ring
+
+/-!
+# Coprime elements of a ring
+
+## Main definitions
+
+* `IsCoprime x y`: that `x` and `y` are coprime, defined to be the existence of `a` and `b` such
+that `a * x + b * y = 1`. Note that elements with no common divisors are not necessarily coprime,
+e.g., the multivariate polynomials `x₁` and `x₂` are not coprime.
+
+See also `RingTheory.Coprime.Lemmas` for further development of coprime elements.
+-/
+
+
+open Classical
+
+universe u v
+
+section CommSemiring
+
+variable {R : Type u} [CommSemiring R] (x y z : R)
+
+/-- The proposition that `x` and `y` are coprime, defined to be the existence of `a` and `b` such
+that `a * x + b * y = 1`. Note that elements with no common divisors are not necessarily coprime,
+e.g., the multivariate polynomials `x₁` and `x₂` are not coprime. -/
+@[simp]
+def IsCoprime : Prop :=
+  ∃ a b, a * x + b * y = 1
+#align is_coprime IsCoprime
+
+variable {x y z}
+
+theorem IsCoprime.symm (H : IsCoprime x y) : IsCoprime y x :=
+  let ⟨a, b, H⟩ := H
+  ⟨b, a, by rw [add_comm, H]⟩
+#align is_coprime.symm IsCoprime.symm
+
+theorem isCoprime_comm : IsCoprime x y ↔ IsCoprime y x :=
+  ⟨IsCoprime.symm, IsCoprime.symm⟩
+#align is_coprime_comm isCoprime_comm
+
+theorem isCoprime_self : IsCoprime x x ↔ IsUnit x :=
+  ⟨fun ⟨a, b, h⟩ => isUnit_of_mul_eq_one x (a + b) <| by rwa [mul_comm, add_mul], fun h =>
+    let ⟨b, hb⟩ := isUnit_iff_exists_inv'.1 h
+    ⟨b, 0, by rwa [zero_mul, add_zero]⟩⟩
+#align is_coprime_self isCoprime_self
+
+theorem isCoprime_zero_left : IsCoprime 0 x ↔ IsUnit x :=
+  ⟨fun ⟨a, b, H⟩ => isUnit_of_mul_eq_one x b <| by rwa [mul_zero, zero_add, mul_comm] at H, fun H =>
+    let ⟨b, hb⟩ := isUnit_iff_exists_inv'.1 H
+    ⟨1, b, by rwa [one_mul, zero_add]⟩⟩
+#align is_coprime_zero_left isCoprime_zero_left
+
+theorem isCoprime_zero_right : IsCoprime x 0 ↔ IsUnit x :=
+  isCoprime_comm.trans isCoprime_zero_left
+#align is_coprime_zero_right isCoprime_zero_right
+
+theorem not_coprime_zero_zero [Nontrivial R] : ¬IsCoprime (0 : R) 0 :=
```

Comment:
```suggestion
theorem not_isCoprime_zero_zero [Nontrivial R] : ¬IsCoprime (0 : R) 0 :=
```

Reply:
Also a few below. These aren't essential, obviously, but this is a cheap moment to rename a few things.